### ISSUE TITLE
Too little steps for training

### ISSUE THREAD

**Issue by eix128** [rhasspy/piper#393]
Hi , when i try to train my speech data.
I saw each epoch is making so much less steps.
And training is not utilizing my gpu much.
It stops after so little steps for each epoch like  

When i look at these trained files
https://huggingface.co/datasets/rhasspy/piper-checkpoints/tree/main/tr/tr_TR/fahrettin/medium

They stop at
 epoch=8099-step=1497980.ckpt

epoch 8099 step 1497980

But my steps too slow
and when i look nvidia-smi
sometimes app uses %90 of gpu power and sometimes stops at %2
How can i utilize gpu fully ?

I use these commands

pretrain cmd args:
--dataset-dir /home/kadirbasol/piperTrain/  --accelerator gpu --devices 1 --batch-size 32 --max-phoneme-ids 400 --max_epochs 19000 --precision 32
train cmd args:
--language tr --single-speaker --input-dir /media/kadirbasol/KadirBASOL/Files/Works/JavaTTS/SoundWorks/piperTrain --output-dir /home/kadirbasol/piperTrain --max-workers 16 --dataset-format ljspeech --single-speaker --sample-rate 22050

I also tried :
--checkpoint-epochs 5
but  when i set it . i cant see any logs like error level etc?

Here is my training:

/home/kadirbasol/repo/piper/src/python/.venv/bin/python /home/kadirbasol/repo/piper/src/python/piper_train/__main__.py --dataset-dir /home/kadirbasol/piperTrain/ --accelerator gpu --devices 1 --batch-size 32 --max-phoneme-ids 400 --max_epochs 19000 --precision 32 
DEBUG:root:Namespace(dataset_dir='/home/kadirbasol/piperTrain/', checkpoint_epochs=None, quality='medium', resume_from_single_speaker_checkpoint=None, logger=True, enable_checkpointing=True, default_root_dir=None, gradient_clip_val=None, gradient_clip_algorithm=None, num_nodes=1, num_processes=None, devices='1', gpus=None, auto_select_gpus=False, tpu_cores=None, ipus=None, enable_progress_bar=True, overfit_batches=0.0, track_grad_norm=-1, check_val_every_n_epoch=1, fast_dev_run=False, accumulate_grad_batches=None, max_epochs=19000, min_epochs=None, max_steps=-1, min_steps=None, max_time=None, limit_train_batches=None, limit_val_batches=None, limit_test_batches=None, limit_predict_batches=None, val_check_interval=None, log_every_n_steps=50, accelerator='gpu', strategy=None, sync_batchnorm=False, precision=32, enable_model_summary=True, weights_save_path=None, num_sanity_val_steps=2, resume_from_checkpoint=None, profiler=None, benchmark=None, deterministic=None, reload_dataloaders_every_n_epochs=0, auto_lr_find=False, replace_sampler_ddp=True, detect_anomaly=False, auto_scale_batch_size=False, plugins=None, amp_backend='native', amp_level=None, move_metrics_to_cpu=False, multiple_trainloader_mode='max_size_cycle', batch_size=32, validation_split=0.1, num_test_examples=5, max_phoneme_ids=400, hidden_channels=192, inter_channels=192, filter_channels=768, n_layers=6, n_heads=2, seed=1234)
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
DEBUG:vits.dataset:Loading dataset: /home/kadirbasol/piperTrain/dataset.jsonl
WARNING:vits.dataset:Skipped 33 utterance(s)
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name    | Type                     | Params
-----------------------------------------------------
0 | model_g | SynthesizerTrn           | 23.7 M
1 | model_d | MultiPeriodDiscriminator | 46.7 M
-----------------------------------------------------
70.4 M    Trainable params
0         Non-trainable params
70.4 M    Total params
281.644   Total estimated model params size (MB)
DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/hparams.yaml
Sanity Checking: 0it [00:00, ?it/s]/home/kadirbasol/repo/piper/src/python/.venv/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:236: PossibleUserWarning: The dataloader, val_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 64 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
Sanity Checking DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]/home/kadirbasol/repo/piper/src/python/.venv/lib/python3.10/site-packages/pytorch_lightning/utilities/data.py:98: UserWarning: Trying to infer the `batch_size` from an ambiguous collection. The batch size we found is 4. To avoid any miscalculations, use `self.log(..., batch_size=batch_size)`.
  warning_cache.warn(
Epoch 0:   0%|          | 0/3 [00:00<?, ?it/s] /home/kadirbasol/repo/piper/src/python/.venv/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:236: PossibleUserWarning: The dataloader, train_dataloader, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 64 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
/home/kadirbasol/repo/piper/src/python/.venv/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py:1892: PossibleUserWarning: The number of training batches (2) is smaller than the logging interval Trainer(log_every_n_steps=50). Set a lower value for log_every_n_steps if you want to see logs for the training epoch.
  rank_zero_warn(
Epoch 0:  67%|██████▋   | 2/3 [00:07<00:03,  3.68s/it, loss=118, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 0: 100%|██████████| 3/3 [00:08<00:00,  2.78s/it, loss=118, v_num=8]
Epoch 0: 100%|██████████| 3/3 [00:08<00:00,  2.78s/it, loss=118, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=0-step=4.ckpt
Epoch 1:  67%|██████▋   | 2/3 [00:01<00:00,  1.35it/s, loss=93.1, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 1: 100%|██████████| 3/3 [00:02<00:00,  1.21it/s, loss=93.1, v_num=8]
Epoch 1: 100%|██████████| 3/3 [00:02<00:00,  1.21it/s, loss=93.1, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=1-step=8.ckpt
Epoch 2:  67%|██████▋   | 2/3 [00:01<00:00,  1.31it/s, loss=81.3, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 2: 100%|██████████| 3/3 [00:02<00:00,  1.13it/s, loss=81.3, v_num=8]
Epoch 2: 100%|██████████| 3/3 [00:02<00:00,  1.13it/s, loss=81.3, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=2-step=12.ckpt
Epoch 3:  67%|██████▋   | 2/3 [00:01<00:00,  1.22it/s, loss=73.3, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 3: 100%|██████████| 3/3 [00:02<00:00,  1.10it/s, loss=73.3, v_num=8]
Epoch 3: 100%|██████████| 3/3 [00:02<00:00,  1.10it/s, loss=73.3, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=3-step=16.ckpt
Epoch 4:  67%|██████▋   | 2/3 [00:01<00:00,  1.29it/s, loss=70.1, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 4: 100%|██████████| 3/3 [00:02<00:00,  1.21it/s, loss=70.1, v_num=8]
Epoch 4: 100%|██████████| 3/3 [00:02<00:00,  1.21it/s, loss=70.1, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=4-step=20.ckpt
Epoch 5:  67%|██████▋   | 2/3 [00:01<00:00,  1.38it/s, loss=57.5, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 5: 100%|██████████| 3/3 [00:02<00:00,  1.22it/s, loss=57.5, v_num=8]
Epoch 5: 100%|██████████| 3/3 [00:02<00:00,  1.21it/s, loss=57.5, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=5-step=24.ckpt
Epoch 6:  67%|██████▋   | 2/3 [00:01<00:00,  1.36it/s, loss=51.8, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 6: 100%|██████████| 3/3 [00:02<00:00,  1.20it/s, loss=51.8, v_num=8]
Epoch 6: 100%|██████████| 3/3 [00:02<00:00,  1.20it/s, loss=51.8, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=6-step=28.ckpt
Epoch 7:  67%|██████▋   | 2/3 [00:01<00:00,  1.24it/s, loss=50.8, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 7: 100%|██████████| 3/3 [00:02<00:00,  1.05it/s, loss=50.8, v_num=8]
Epoch 7: 100%|██████████| 3/3 [00:02<00:00,  1.05it/s, loss=50.8, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=7-step=32.ckpt
Epoch 8:  67%|██████▋   | 2/3 [00:01<00:00,  1.32it/s, loss=49.2, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 8: 100%|██████████| 3/3 [00:02<00:00,  1.06it/s, loss=49.2, v_num=8]
Epoch 8: 100%|██████████| 3/3 [00:02<00:00,  1.06it/s, loss=49.2, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=8-step=36.ckpt
Epoch 9:  67%|██████▋   | 2/3 [00:01<00:00,  1.35it/s, loss=46.7, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 9: 100%|██████████| 3/3 [00:02<00:00,  1.12it/s, loss=46.7, v_num=8]
Epoch 9: 100%|██████████| 3/3 [00:02<00:00,  1.12it/s, loss=46.7, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=9-step=40.ckpt
Epoch 10:  67%|██████▋   | 2/3 [00:01<00:00,  1.24it/s, loss=43.8, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 10: 100%|██████████| 3/3 [00:02<00:00,  1.05it/s, loss=43.8, v_num=8]
Epoch 10: 100%|██████████| 3/3 [00:02<00:00,  1.05it/s, loss=43.8, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=10-step=44.ckpt
Epoch 11:  67%|██████▋   | 2/3 [00:01<00:00,  1.39it/s, loss=43, v_num=8]  
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 11: 100%|██████████| 3/3 [00:02<00:00,  1.07it/s, loss=43, v_num=8]
Epoch 11: 100%|██████████| 3/3 [00:02<00:00,  1.07it/s, loss=43, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=11-step=48.ckpt
Epoch 12:  67%|██████▋   | 2/3 [00:01<00:00,  1.36it/s, loss=39.9, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 12: 100%|██████████| 3/3 [00:02<00:00,  1.07it/s, loss=39.9, v_num=8]
Epoch 12: 100%|██████████| 3/3 [00:02<00:00,  1.07it/s, loss=39.9, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=12-step=52.ckpt
Epoch 13:  67%|██████▋   | 2/3 [00:01<00:00,  1.39it/s, loss=40.5, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 13: 100%|██████████| 3/3 [00:02<00:00,  1.18it/s, loss=40.5, v_num=8]
Epoch 13: 100%|██████████| 3/3 [00:02<00:00,  1.18it/s, loss=40.5, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=13-step=56.ckpt
Epoch 14:  67%|██████▋   | 2/3 [00:01<00:00,  1.34it/s, loss=39.6, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 14: 100%|██████████| 3/3 [00:02<00:00,  1.16it/s, loss=39.6, v_num=8]
Epoch 14: 100%|██████████| 3/3 [00:02<00:00,  1.16it/s, loss=39.6, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=14-step=60.ckpt
Epoch 15:  67%|██████▋   | 2/3 [00:01<00:00,  1.33it/s, loss=37.8, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Epoch 15: 100%|██████████| 3/3 [00:02<00:00,  1.09it/s, loss=37.8, v_num=8]
Epoch 15: 100%|██████████| 3/3 [00:02<00:00,  1.09it/s, loss=37.8, v_num=8]DEBUG:fsspec.local:open file: /home/kadirbasol/piperTrain/lightning_logs/version_8/checkpoints/epoch=15-step=64.ckpt
Epoch 16:  67%|██████▋   | 2/3 [00:01<00:00,  1.32it/s, loss=38.3, v_num=8]
Validation: 0it [00:00, ?it/s]
Validation:   0%|          | 0/1 [00:00<?, ?it/s]
Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]

**Comment by eix128** [comment]
Okay i have increased gpu utilization by just copy paste same dataset 10x more.
i have 100 sample of text and wav now.
how much do i need for training?

btw , my training stuck between loss=16 and loss=15
how to go down more? do i need more speech data ?
100 is not enough ?

**Comment by nmstoker** [comment]
Hello @eix128 - it's a while since you posted but if you're still interested in this, I had a couple of comments (but I'm no Piper expert!)

* I noticed that your log output suggests it skipped a number of utterances:
  `Skipped 33 utterance(s)`

  Now I'm not quite sure how many utterances you had to start with but if it's not many and it removed a significant portion then that likely explains things 

  I can't tell for sure but I wonder if it's because they are over the max phonemes limit you set (ie 400)

* To answer your question regarding how many utterances you need, I haven't explored the lower bounds but my experience with similar tools would suggest it's good to have at least a few thousand, and crucially you need them to have good coverage of all the phonemes used in the language you're training your model for

  With Piper I got decent results with a private dataset I produced with ~25k diverse utterances but I suspect that's more than strictly necessary!

  As well as coverage, you generally want them to be fairly consistent in tone / speaking style

Hope that helps a little?

If you tried Piper out further (now or earlier), do post an update for how you got on, what worked / didn't, as it may help others!
